scalameta/semantic/shared/src/main/scala/scala/meta/semantic/Denotation.scala

```diff
@@ -0,0 +1,110 @@
+package scala.meta
+package semantic
+
+import org.scalameta.adt._
+import org.scalameta.data._
+
+@data class Denotation(flags: Long, name: String, info: String) extends HasFlags {
```

Comment:
To quote `Denotations.scala` from Dotty (https://github.com/lampepfl/dotty/blob/fa13f8d772cc53a8dce5e6854309f00288876d0d/compiler/src/dotty/tools/dotc/core/Denotations.scala#L56-L58):

>  *  Symbol                   A label for a definition or declaration in one compiler run
> *  Denotation              The meaning of a named type or symbol during a period

This corresponds exactly to what we're doing in scalameta. A symbol is just a lightweight wrapper of a string id, and a denotation represents its meaning within a given mirror (just like in Dotty, a symbol can have multiple denotations depending on the current context).

Even the fields of the `Denotation` class in scalameta are inspired by the fields of a similar class in Dotty: `flags`, `name`, `info`. In the future, we plan to add `privateWithin` and most likely `owner`.

All in all, I'm confident that separating `Symbol.id` from `Symbol.info` and friends is a very important insight. In scalac, these concepts are mixed in a fat `Symbol` object, which makes it hard to reliably metaprogram between compiler runs. @densh and I started realizing this in late 2012 when working on hygiene from scala.reflect (back then scalameta didn't exist yet), and I'm happy that in 2017 this is something that's employed in multiple metaprogramming frameworks.